File: src/poke/client/comm/CommConnection.java
/*
 * copyright 2014, gash
 * 
 * Gash licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
package poke.client.comm;

import io.netty.bootstrap.Bootstrap;
import io.netty.channel.Channel;
import io.netty.channel.ChannelFuture;
import io.netty.channel.ChannelFutureListener;
import io.netty.channel.ChannelOption;
import io.netty.channel.ChannelPipeline;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioSocketChannel;
import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
import io.netty.handler.codec.LengthFieldPrepender;
import io.netty.handler.codec.protobuf.ProtobufDecoder;
import io.netty.handler.codec.protobuf.ProtobufEncoder;

import java.util.concurrent.LinkedBlockingDeque;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import poke.server.ServerHandler;

import com.google.protobuf.GeneratedMessage;

import eye.Comm.Request;


public class CommConnection {
	protected static Logger logger = LoggerFactory.getLogger("connect");

	private String host;
	private int port;
	public ChannelFuture channel; // do not use directly call connect()!
	private EventLoopGroup group;
	private CommHandler handler;
	
	// our surge protection using a in-memory cache for messages
	private LinkedBlockingDeque<com.google.protobuf.GeneratedMessage> outbound;

	// message processing is delegated to a threading model
	private OutboundWorker worker;


	public CommConnection(String host, int port) {
		this.host = host;
		this.port = port;

		init();
	}

	public void release() {
		group.shutdownGracefully();
	}


	public void sendMessage(Request req) throws Exception {
		// enqueue message
		outbound.put(req);
	}

	public void addListener(CommListener listener) {
		// note: the handler should not be null as we create it on construction

		try {
			handler.addListener(listener);
		} catch (Exception e) {
			logger.error("failed to add listener", e);
		}
	}

	private void init() {
		// the queue to support client-side surging
		outbound = new LinkedBlockingDeque<com.google.protobuf.GeneratedMessage>();

		group = new NioEventLoopGroup();
		try {
			handler = new CommHandler();
			Bootstrap b = new Bootstrap();
			b.group(group).channel(NioSocketChannel.class).handler(handler);
			b.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 10000);
			b.option(ChannelOption.TCP_NODELAY, true);
			b.option(ChannelOption.SO_KEEPALIVE, true);

			// Make the connection attempt.
			channel = b.connect(host, port).syncUninterruptibly();
			channel.awaitUninterruptibly(5000l);
			if(channel == null)
				System.out.println("Channel is null, not able to connect to Host: "+host+"  Port: "+port);
			else
				System.out.println("Channel created, not null");

			//setting pipeline parameters
			ChannelPipeline pipeline = channel.channel().pipeline();
			pipeline.addLast("frameDecoder", new LengthFieldBasedFrameDecoder(67108864, 0, 4, 0, 4));
			pipeline.addLast("protobufDecoder", new ProtobufDecoder(eye.Comm.Request.getDefaultInstance()));
			pipeline.addLast("frameEncoder", new LengthFieldPrepender(4));
			pipeline.addLast("protobufEncoder", new ProtobufEncoder());
			pipeline.addLast("handler", handler);

			
			// want to monitor the connection to the server s.t. if we loose the
			// connection, we can try to re-establish it.
			ClientClosedListener ccl = new ClientClosedListener(this);
			handler.setChannel(channel.channel());
			channel.channel().closeFuture().addListener(ccl);

		} catch (Exception ex) {
			logger.error("failed to initialize the client connection", ex);

		}

		// start outbound message processor
		worker = new OutboundWorker(this);
		worker.start();
	}

	protected Channel connect() {
		// Start the connection attempt.
		if (channel == null) {
			init();
		}

		if (channel.isDone() && channel.isSuccess())
		{
			System.out.println("Channel is success");
			return channel.channel();
		}
		else
			throw new RuntimeException("Not able to establish connection to server");
	}


	protected class OutboundWorker extends Thread {
		CommConnection conn;
		boolean forever = true;

		public OutboundWorker(CommConnection conn) {
			this.conn = conn;

			if (conn.outbound == null)
				throw new RuntimeException("connection worker detected null queue");
		}

		@Override
		public void run() {
			Channel ch = conn.connect();
			if (ch == null || !ch.isOpen()) {
				CommConnection.logger.error("connection missing, no outbound communication");
				return;
			}
			while (true) {
				if (!forever && conn.outbound.size() == 0)
					break;

				try {
					// block until a message is enqueued
					GeneratedMessage msg = conn.outbound.take();
					//System.out.println("Message received");
					if (ch.isWritable()) {
						CommHandler handler = conn.connect().pipeline().get(CommHandler.class);
						handler.setChannel(ch);
						ch.writeAndFlush(msg);
					} else
						conn.outbound.putFirst(msg);
				} catch (InterruptedException ie) {
					break;
				} catch (Exception e) {
					CommConnection.logger.error("Unexpected communcation failure", e);
					break;
				}
			}

			if (!forever) {
				CommConnection.logger.info("connection queue closing");
			}
		}
	}

	public static class ClientClosedListener implements ChannelFutureListener {
		CommConnection cc;

		public ClientClosedListener(CommConnection cc) {
			this.cc = cc;
		}

		@Override
		public void operationComplete(ChannelFuture future) throws Exception {
			// we lost the connection or have shutdown.
			System.out.println("Channel closed");
			// @TODO if lost, try to re-establish the connection
		}
	}
}

File: python/src/CommConnection.py
import sys

sys.path.append("netty-all-4.0.15.Final.jar")
sys.path.append("protobuf-java-2.5.0.jar")

from eye.Comm import Request, Header, Payload, RoutingPath, Ping, JobOperation
from com.google.protobuf import GeneratedMessage
from io.netty.bootstrap import Bootstrap, ChannelFactory
from io.netty.buffer import PooledByteBufAllocator, Unpooled
from io.netty.channel.nio import NioEventLoopGroup
from io.netty.channel.socket.nio import NioSocketChannel
from io.netty.channel import Channel, ChannelPipeline, SimpleChannelInboundHandler, ChannelOption
from io.netty.handler.codec import LengthFieldBasedFrameDecoder, LengthFieldPrepender
from io.netty.handler.codec.protobuf import ProtobufDecoder, ProtobufEncoder
from CommHandler import CommHandler

class CommConnection():
    
    def __init__(self, host, port):
        self.host = host
        self.port = port
        
        group = NioEventLoopGroup()
        try:
            #Create the eventLoopGroup and the channel
            self.handler = CommHandler()
            bootstrap = Bootstrap().group(group).channel(NioSocketChannel).handler(self.handler)
            bootstrap.option(ChannelOption.CONNECT_TIMEOUT_MILLIS,10000)
            bootstrap.option(ChannelOption.TCP_NODELAY, True)
            bootstrap.option(ChannelOption.SO_KEEPALIVE, True) 
            #bootstrap.option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
            channel = bootstrap.connect(host, port).syncUninterruptibly()
            channel.awaitUninterruptibly(2000)
            if channel is None:
                print "Could not connect to the Server"
            else:
                print "Channel created" 
            
            pipeline = channel.channel().pipeline()
            pipeline.addLast("frameDecoder", LengthFieldBasedFrameDecoder(67108864, 0, 4, 0, 4))
            pipeline.addLast("protobufDecoder", ProtobufDecoder(Request.getDefaultInstance()))
            pipeline.addLast("frameEncoder", LengthFieldPrepender(4))
            pipeline.addLast("protobufEncoder", ProtobufEncoder())    
            pipeline.addLast("handler", CommHandler())
            
            self.handler.setChannel(channel.channel())
        except:
            print sys.exc_info()[0]
        finally:
            print "finally!!"
            #group.shutdownGracefully()
    
    def poke(self):
        ping = Ping.newBuilder()
        ping.setTag("test poke")
        ping.setNumber(5)
        
        #Payload
        r = Request.newBuilder()
        p = Payload.newBuilder()
        p.setPing(ping.build())
        r.setBody(p.build())

        #header with routing info
        h = Header.newBuilder()
        h.setOriginator("client")
        h.setTag("test finger")
        h.setRoutingId(Header.Routing.PING)
        r.setHeader(h.build())

        req = r.build()
        self.handler.send(req)
        
    def jobrequest(self):
        ping = Ping.newBuilder()
        ping.setTag("test job")
        ping.setNumber(5)
        
        jobOp = JobOperation.newBuilder()
        jobOp.setAction(JobOperation.JobAction.LISTJOBS)
        
        #Payload
        r = Request.newBuilder()
        p = Payload.newBuilder()
        p.setPing(ping.build())
        p.setJobOp(jobOp.build())
        r.setBody(p.build())

        #header with routing info
        h = Header.newBuilder()
        h.setOriginator("client")
        h.setTag("test finger")
        h.setRoutingId(Header.Routing.JOBS)
        r.setHeader(h.build())

        req = r.build()
        self.handler.send(req)    
                    
